Build PDF files using latexmk

humitos opened the pull request:
This PR makes [a Japanese project](https://github.com/TECS-WG/tecs-docs/issues/7) to build properly. It would be awesome to have more projects as examples to test here. Although, since this code is behind a feature flag we can merge without worry too much for now on making it work perfectly.

I'm referencing these issues/prs to be closed here since most of the requests are covered in this PR. It may be some things missings, like supporting `xindy` or using environment variables like `XINDYOPTS`.


Closes #1556
Closes #4454
Closes #5405

stsewd commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -395,6 +397,57 @@ def build(self):
+        # ``latex_engine`` is ``platex``
+        pdfs = []
+        if self.project.language == 'ja':
+            pdfs = latex_path.glob('*.pdf')
```
Why is this only required by ja lang? Don't we already run this whole function under a feature flag?

humitos replied:
Japanese language is the only one that requires this extra step. I don't know _exactly_ why but most of the documentation that I read differentiate this language from the others. I suppose it's because it mix kanji (Chinese) with its own symbols.

I took this step from the Makefile generated by Sphinx when the language is Japanese.

jfbu replied:
The Japanese language uses `latex+dvipdfmx` path. Until TeXLive 2015 release, `extractbb` step was needed. But it is not needed with newer `dvipdfmx` binary. Here is a quote of an [exchange](https://tug.org/pipermail/tex-live/2017-January/039545.html) I had in 2017 on TeXLive mailing list:

> > > Can someone confirm that since TL2014 there is no need
> > > on Unix-like systems to prepare beforehand .xbb
> > > files for graphics inclusion of .png, .jpeg, ..., 
> > > when doing latex+dvipdfmx or platex+dvipdfmx ?
> > > (and that it was either needed before, or at least
> > > shell-escape had to be enabled) ?
> > > 
> > 
> 
> > I remember xbb files were necessary for TL2014 or earlier.
> > Japanese people has been heavily using dvipdfmx, so they added
> > extractbb to their local texmf.cnf by hand
> > 
> > ...
> 

> 
> Thanks for the information ! This is for support
> of some other software, now I know they can avoid
> the explicit extractbb calls but under assumption
> of a TL2015 or more recent install,

Thus Sphinx may soon drop the `extractbb` and make the Japanese set-up in Makefile less singular: indexe 2.0 release requires "LaTeX builder now depends on TeX Live 2015 or above." but we forgot to actually remove the `extractbb` dependency as we don't want either to break too easily user projects with older TeX installations.

jfbu replied:
I have prepared https://github.com/sphinx-doc/sphinx/pull/6189 for Sphinx to remove the extra handling of image files for Japanese language by `extractbb` as it is superfluous for TL2015 or later based TeX distribution. Not sure it will be merged during 2.x series.

stsewd commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -395,6 +397,57 @@ def build(self):
+        if self.project.language == 'ja':
+            pdfs = latex_path.glob('*.pdf')
+
+        for image in itertools.chain(images, pdfs):
```
image and pdfs are not related I guess, but not sure about this step

stsewd commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -395,6 +397,57 @@ def build(self):
+            '-r',
+            rcfile,
+
+            # FIXME: check for platex here as well
```
I'm confused about platex and latexmk

humitos replied:
`latexmk` uses internally `pdflatex` or `xelatex` or `lualatex` or `platex` (defined by [`latex_engine`](http://www.sphinx-doc.org/en/master/usage/configuration.html#confval-latex_engine)

stsewd commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -395,6 +397,57 @@ def build(self):
+            cwd=latex_cwd,
+        )
+
+        self.pdf_file_name = f'{self.project.slug}.pdf'
```
Where did this come from? From the `-jobname` option?

humitos replied:
Yes.

humitos commented:
> This feature flag should be documented so users that need it can request it to us

Yes. I don't want to expose this _yet_ since we don't have too much working examples. Although, I plan to write a guide on our docs explaining what are the required steps from the users to build PDF for these languages.

humitos commented:
> It would be awesome to have more projects as examples to test here

I tested some projects today and I was able to build most of them. Some of them did not build because other problems not related to this PR (missing png file, invalid rst table on the docs, or similar)

humitos commented:
> This feature flag should be documented so users that need it can request it to us

I added this since I feel more comfortable with these code after some testing and also because I added a default configuration in our `conf.py.tmpl` which does not need user intervention.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -395,6 +400,63 @@ def build(self):
+            'cat',
+            rcfile,
+            cwd=latex_cwd,
+        )
```
Is this to output to the user, or some other reason?

humitos replied:
Yes. This is only to show the content of the `latexmkrc` file to the user.

I want to show this because that file is built by Sphinx depending on some configurations. Also, this is the file that will say what command execute to build the PDF in the end. So, without this output will be very hard to debug a problem.

humitos commented:
> Seems reasonable. Is the goal to turn this on for all Chinese & Japanese languages by default after we test it behind the feature flag?

I'd say yes. Although, not only for Chinese & Japanese, but for all the languages. This is the default way of building PDF on Sphinx >=1.6 and seems to be more robust that the one that we are using. Also, it allows customization using Sphinx common configuration options.

> Does this require additional ops or docker changes, to ensure we have the proper binaries installed?

No. It shouldn't.

humitos commented on readthedocs/doc_builder/templates/doc_builder/conf.py.tmpl:
```diff
@@ -140,3 +140,38 @@ if 'extensions' in globals():
+    project_language in ('zh_CN', 'zh_TW'),
+])
+
+japanase = any([
```
Yes. It's a typo!

humitos commented:
@jfbu thanks for your feedback on this!

I don't have too much experience with LaTeX and building other languages that are not Spanish or English. So, I decided to follow what current stable Sphinx version was doing instead of "propose something better/polished" to avoid entering new and different bugs to the ones that Sphinx already have.

Once Sphinx 2.x is released, we could refactor these steps and make them work better but for now I prefer to keep it stable.

From my tests locally I can say that I was able to build PDF in Chinese, Japanese and English without problem. So, this is way better to what we had.

Let's see how it goes in production with real projects and we can iterate from there.
